compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -206,9 +206,16 @@ impl<'s> AllocDecodingSession<'s> {
             (alloc_kind, decoder.position())
         });
 
+        // We are going to hold this lock during the entire decoding of this allocation, which may
+        // require that we decode other allocations. This cannot deadlock for two reasons:
+        // It is impossible to create an allocation that contains a pointer to itself and thus
+        // deadlock with a single thread, because attempting to create such a pointer immediately
+        // blows the stack.
+        // It is also impossible to create two allocations (call them A and B) where A is a pointer to B, and B
+        // is a pointer to A, because const eval itself rejects such cycles.
```

Comment:
It doesn't reject them, it will cause a query cycle because `eval(A) -> eval(B) -> eval(A)`, but that's just wording nitpicking